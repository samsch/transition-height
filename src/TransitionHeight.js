import React from 'react';

// non-'visible' overflow style to be added to inner wrapper, which stops
// margin collapsing out from children.
const styleOverFlowAuto = {
  overflowY: 'auto',
};

class TransitionHeight extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      height: undefined,
    };
    this.getWrapperDiv = div => {
      this.divEl = div;
    };
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.trigger !== this.props.trigger) {
      this.setState({
        height: this.divEl.offsetHeight,
      });
    }
  }
  componentDidUpdate(prevProps) {
    if (this.state.height != null && prevProps.trigger !== this.props.trigger) {
      this.setState(
        {
          height: this.divEl.offsetHeight,
        },
        () => {
          setTimeout(() => {
            this.setState({
              height: undefined,
            });
          }, this.props.transitionTimeSeconds * 1000);
        }
      );
    }
  }
  render() {
    const outerWrapperStyle = {
      transition: 'height ' + this.props.transitionTimeSeconds + 's ease',
      height: this.state.height,
      overflowY: this.state.height == null ? 'auto' : 'hidden',
    };
    return (
      <div style={outerWrapperStyle}>
        <div style={styleOverFlowAuto} ref={this.getWrapperDiv}>
          {this.props.children}
        </div>
      </div>
    );
  }
}
TransitionHeight.defaultProps = {
  transitionTimeSeconds: 0.2,
};
export default TransitionHeight;
