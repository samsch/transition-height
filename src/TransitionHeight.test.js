/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import TransitionHeight from './TransitionHeight';

describe('TransitionHeight', () => {
  test('initially renders', () => {
    const component = renderer.create(
      <TransitionHeight>
        <span>Content</span>
      </TransitionHeight>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  // Needs more test, but not as trivial to write, since the component is
  // dependant on the DOM and lifecycle methods.
});
