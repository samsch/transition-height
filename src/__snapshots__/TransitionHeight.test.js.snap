// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionHeight initially renders 1`] = `
<div
  style={
    Object {
      "height": undefined,
      "overflowY": "auto",
      "transition": "height 0.2s ease",
    }
  }
>
  <div
    style={
      Object {
        "overflowY": "auto",
      }
    }
  >
    <span>
      Content
    </span>
  </div>
</div>
`;
